lib/Sema/CSGen.cpp

```diff
@@ -3431,6 +3431,17 @@ class InferUnresolvedMemberConstraintGenerator : public ConstraintGenerator {
     return Expr->getType();
   }
 
+  Type visitDeclRefExpr(DeclRefExpr *Expr) override {
+    if (Target != Expr) {
+      if (Type Ty = Expr->getType())
+        return Ty;
+      return ConstraintGenerator::visitDeclRefExpr(Expr);
+    }
+    // Otherwise, create a type variable saying we know nothing about this expr.
```

Comment:
Instead of having a subclass for this purpose it could be a flag for the constraint generator